### PR TITLE
updating balance calculations

### PR DESCRIPTION
**Pull request by aupiff**
- single and two-party balances will now exclude pending bilateral settlement
credits

@canterberry there's some ugly but fairly straightforward sql below. If you have any suggestions as to how to simplify this, let me know. I'm going to ponder this myself for a few minutes...

### CODE REVIEW

File: lndr-backend/src/Lndr/Db/VerifiedCredits.hs [hunk 1]
```diff
@@ -73,15 +73,19 @@ verifyCreditByHash :: Text -> Connection -> IO Int
 verifyCreditByHash hash conn = fromIntegral <$> execute conn "UPDATE settlements SET verified = TRUE WHERE hash = ?" (Only hash)
 
 
+-- Balance is calculated per ucac and only includes bilateral non-settlement
+-- credits and transaction-verified bilateral settlementcredits
 userBalance :: Address -> Address -> Connection -> IO Integer
 userBalance addr ucac conn = do
-    [Only balance] <- query conn "SELECT (SELECT COALESCE(SUM(amount), 0) FROM verified_credits WHERE creditor = ? AND ucac = ?) - (SELECT COALESCE(SUM(amount), 0) FROM verified_credits WHERE debtor = ? AND ucac = ?)" (addr, ucac, addr, ucac) :: IO [Only Scientific]
+    [Only balance] <- query conn "SELECT (SELECT COALESCE(SUM(verified_credits.amount), 0) FROM verified_credits LEFT JOIN settlements USING(hash) WHERE creditor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE) - (SELECT COALESCE(SUM(verified_credits.amount), 0) FROM verified_credits LEFT JOIN settlements USING (hash) WHERE debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE)" (addr, ucac, addr, ucac) :: IO [Only Scientific]
```
Comment (canterberry) [inline lndr-backend/src/Lndr/Db/VerifiedCredits.hs]: In plain English, what is the intention of this query? Is there any way to split it across multiple lines for readability?

Comment (canterberry) [reply lndr-backend/src/Lndr/Db/VerifiedCredits.hs]: Is this something that could potentially be more legibly described in Haskell from the results of a more straightforward query?

Comment (canterberry) [reply lndr-backend/src/Lndr/Db/VerifiedCredits.hs]:
```sql
SELECT (
  SELECT
    COALESCE(
      SUM(verified_credits.amount),
      0
    )
  FROM
    verified_credits
      LEFT JOIN
    settlements
      USING(hash)
  WHERE
    creditor = ?
      AND
    ucac = ?
      AND
    verified IS DISTINCT FROM FALSE
) - (
  SELECT
    COALESCE(
      SUM(verified_credits.amount),
      0
    )
  FROM
    verified_credits
      LEFT JOIN
    settlements
      USING (hash)
  WHERE
    debtor = ?
      AND
    ucac = ?
      AND
    verified IS DISTINCT FROM FALSE
)
```

File: lndr-backend/src/Lndr/Db/VerifiedCredits.hs [hunk 2]
```diff
@@ -73,15 +73,19 @@ verifyCreditByHash :: Text -> Connection -> IO Int
 verifyCreditByHash hash conn = fromIntegral <$> execute conn "UPDATE settlements SET verified = TRUE WHERE hash = ?" (Only hash)
 
 
+-- Balance is calculated per ucac and only includes bilateral non-settlement
+-- credits and transaction-verified bilateral settlementcredits
```
Comment (canterberry) [inline lndr-backend/src/Lndr/Db/VerifiedCredits.hs]: https://youtu.be/ILysMhm8lXs?t=9s

File: lndr-backend/src/Lndr/Db/VerifiedCredits.hs [hunk 3]
```diff
@@ -73,15 +73,19 @@ verifyCreditByHash :: Text -> Connection -> IO Int
...
     return . floor $ balance
 
 
+-- Two-party balance is calculated per ucac and only includes bilateral
+-- non-settlement credits and transaction-verified bilateral settlementcredits
 twoPartyBalance :: Address -> Address -> Address -> Connection -> IO Integer
 twoPartyBalance addr counterparty ucac conn = do
-    [Only balance] <- query conn "SELECT (SELECT COALESCE(SUM(amount), 0) FROM verified_credits WHERE creditor = ? AND debtor = ? AND ucac = ?) - (SELECT COALESCE(SUM(amount), 0) FROM verified_credits WHERE creditor = ? AND debtor = ? AND ucac = ?)" (addr, counterparty, ucac, counterparty, addr, ucac) :: IO [Only Scientific]
+    [Only balance] <- query conn "SELECT (SELECT COALESCE(SUM(verified_credits.amount), 0) FROM verified_credits LEFT JOIN settlements USING(hash) WHERE creditor = ? AND debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE) - (SELECT COALESCE(SUM(verified_credits.amount), 0) FROM verified_credits LEFT JOIN settlements USING(hash) WHERE creditor = ? AND debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE)" (addr, counterparty, ucac, counterparty, addr, ucac) :: IO [Only Scientific]
```
Comment (canterberry) [inline lndr-backend/src/Lndr/Db/VerifiedCredits.hs]: First pass:

```sql
SELECT (
  SELECT
    COALESCE(SUM(verified_credits.amount), 0)
  FROM
    verified_credits LEFT JOIN settlements USING(hash)
  WHERE
    creditor = ? AND debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE
) - (
  SELECT
    COALESCE(SUM(verified_credits.amount), 0)
  FROM
    verified_credits LEFT JOIN settlements USING(hash)
  WHERE
    creditor = ? AND debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE
)
```

Second pass:

```sql
SELECT
  COALESCE(SUM(verified_credits.amount), 0)
FROM
  verified_credits LEFT JOIN settlements USING(hash)
WHERE
  creditor = ? AND debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE
```

…and…

```sql
SELECT
  COALESCE(SUM(verified_credits.amount), 0)
FROM
  verified_credits LEFT JOIN settlements USING(hash)
WHERE
  creditor = ? AND debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE
```

Similar queries here to `userBalance`, except in this case, the two individual SQL queries are **identical**! Same advice goes for this as for above, then, except here we only need to decompose into one additional function, to calculate the total volume of credit from one user to another. We'd just call that once for one direction, again for the other direction, and in Haskell land, take the difference between the two results.

File: lndr-backend/src/Lndr/Db/VerifiedCredits.hs (hunk 1 above)
Comment (aupiff) [reply lndr-backend/src/Lndr/Db/VerifiedCredits.hs]: I wanted to stay out of haskell land entirely because this is a frequently-made query and shouldn't require two round trips to the DB. Let me glance at the comments you made below now...

File: lndr-backend/src/Lndr/Db/VerifiedCredits.hs (hunk 3 above)
Comment (aupiff) [reply lndr-backend/src/Lndr/Db/VerifiedCredits.hs]: Yeah, I don't think it's worth using two round trips to the DB, but I know this looks better. Originally, the code was exactly as you described.

**Comment by aupiff** [comment]
@canterberry totally agree it enhances readability. I'll make another PR which uses higher-level psql lib that makes things much clearer but results in a single trip to DB. We can decide if we want to switch to that other library then.
